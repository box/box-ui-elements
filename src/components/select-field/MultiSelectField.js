// @flow
import React from 'react';

import BaseSelectField from './BaseSelectField';

type Props = {
    /** Function will be called with an array of all selected options after user selects a new option */
    onChange: Function,
};

const MultiSelectField = ({ ...rest }: Props) => <BaseSelectField {...rest} multiple />;

export default MultiSelectField;
