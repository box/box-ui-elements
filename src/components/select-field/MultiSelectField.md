### Examples

```
initialState = {
    selectedValues: [2, 3],
};

const options = [
    { displayText: 'Option 1', value: 1 },
    { displayText: 'Option 2', value: 2 },
    { displayText: 'Option 3', value: 3 },
];

const handleChange = selectedOptions => {
    setState({
        selectedValues: selectedOptions.map(option => option.value),
    });
};

<MultiSelectField
    onChange={ handleChange }
    options={ options }
    placeholder="Choose something"
    selectedValues={ state.selectedValues }
/>
```

Clear Option Enabled

```
initialState = {
    selectedValues: [2, 3],
};

const options = [
    { displayText: 'Option 1', value: 1 },
    { displayText: 'Option 2', value: 2 },
    { displayText: 'Option 3', value: 3 },
];

const handleChange = selectedOptions => {
    setState({
        selectedValues: selectedOptions.map(option => option.value),
    });
};

<MultiSelectField
    onChange={ handleChange }
    options={ options }
    placeholder="Choose something"
    selectedValues={ state.selectedValues }
    shouldShowClearOption={true}
/>
```

Search Input Enabled

```
initialState = {
    selectedValues: [2, 3],
};

const options = [
    { displayText: 'Option 1', value: 1 },
    { displayText: 'Option 2', value: 2 },
    { displayText: 'Option 3', value: 3 },
];

const handleChange = selectedOptions => {
    setState({
        selectedValues: selectedOptions.map(option => option.value),
    });
};

<MultiSelectField
    onChange={ handleChange }
    options={ options }
    placeholder="Choose something"
    selectedValues={ state.selectedValues }
    shouldShowSearchInput={true}
/>
```

Search Input and Clear Option Enabled

```
initialState = {
    selectedValues: [2, 3],
};

const options = [
    { displayText: 'Option 1', value: 1 },
    { displayText: 'Option 2', value: 2 },
    { displayText: 'Option 3', value: 3 },
];

const handleChange = selectedOptions => {
    setState({
        selectedValues: selectedOptions.map(option => option.value),
    });
};

<MultiSelectField
    onChange={ handleChange }
    options={ options }
    placeholder="Choose something"
    selectedValues={ state.selectedValues }
    shouldShowSearchInput={true}
    shouldShowClearOption={true}
/>
```

Labeled

```
initialState = {
    selectedValues: [2, 3],
};

const options = [
    { displayText: 'Option 1', value: 1 },
    { displayText: 'Option 2', value: 2 },
    { displayText: 'Option 3', value: 3 },
];

const handleChange = selectedOptions => {
    setState({
        selectedValues: selectedOptions.map(option => option.value),
    });
};

<Label text="Multi select field">
    <MultiSelectField
        onChange={ handleChange }
        options={ options }
        placeholder="Choose something"
        selectedValues={ state.selectedValues }
    />
</Label>
```

Header Content

```
initialState = {
    selectedValues: [2, 3],
};

const options = [
    { displayText: 'Option 1', value: 1 },
    { displayText: 'Option 2', value: 2 },
    { displayText: 'Option 3', value: 3 },
];

const handleChange = selectedOptions => {
    setState({
        selectedValues: selectedOptions.map(option => option.value),
    });
};

<MultiSelectField
    headerContent={'Header Title'}
    onChange={ handleChange }
    options={ options }
    placeholder="Choose something"
    selectedValues={ state.selectedValues }
/>
```

Invalid

```
initialState = {
    selectedValues: [2, 3],
};

const options = [
    { displayText: 'Option 1', value: 1 },
    { displayText: 'Option 2', value: 2 },
    { displayText: 'Option 3', value: 3 },
];

const handleChange = selectedOptions => {
    setState({
        selectedValues: selectedOptions.map(option => option.value),
    });
};

<Label text="Multi select field">
    <MultiSelectField
        error="oops"
        onChange={ handleChange }
        options={ options }
        placeholder="Choose something"
        selectedValues={ state.selectedValues }
    />
</Label>
```
