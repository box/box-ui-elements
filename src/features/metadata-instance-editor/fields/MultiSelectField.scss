@import '../../../styles/variables';

.metadata-instance-editor-field-multi-select {
    margin: 0 0 20px;

    .label {
        word-wrap: break-word;
    }

    label .select-container {
        display: block;
        margin-top: 5px;

        .select-button {
            z-index: inherit;
        }
    }

    &-desc {
        color: $sevens;
        word-wrap: break-word;
    }
}
