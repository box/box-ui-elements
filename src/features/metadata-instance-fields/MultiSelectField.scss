@import '../../styles/variables';

.bdl-MultiSelectField {
    margin: 0 0 20px;

    .label {
        word-wrap: break-word;
    }

    label .select-container {
        display: block;

        .select-button {
            z-index: inherit;
        }
    }
}

.bdl-MultiSelectField-desc {
    display: inline-block;
    margin-bottom: 5px;
    color: $bdl-gray-62;
    word-wrap: break-word;
}
