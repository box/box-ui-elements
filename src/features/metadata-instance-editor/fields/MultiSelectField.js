// @flow
import * as React from 'react';
import { FormattedMessage } from 'react-intl';

import Label from 'components/label/Label';
import MultiSelect from 'components/select-field/MultiSelectField';
import type { SelectOptionProp } from 'components/select-field/props';

import type { FieldValue, TemplateFieldOption } from '../flowTypes';
import messages from '../messages';

import './MultiSelectField.scss';

type Props = {
    dataKey: string,
    dataValue?: FieldValue,
    description?: string,
    displayName: string,
    onChange: (key: string, value: FieldValue) => void,
    onRemove: (key: string) => void,
    options?: Array<TemplateFieldOption>,
};

const MultiSelectField = ({
    dataKey,
    dataValue,
    displayName,
    description,
    onChange,
    onRemove,
    options = [],
}: Props) => {
    const placeholder = <FormattedMessage {...messages.metadataFieldMultiSelectValue} />;

    return (
        <div className="metadata-instance-editor-field-multi-select">
            <Label text={displayName}>
                {!!description && <i className="metadata-instance-editor-field-multi-select-desc">{description}</i>}
                <MultiSelect
                    onChange={(selectedOptions: Array<SelectOptionProp>) => {
                        if (selectedOptions.length) {
                            onChange(dataKey, selectedOptions.map(({ value }) => value));
                        } else {
                            onRemove(dataKey);
                        }
                    }}
                    options={options.map(option => ({
                        displayText: option.key,
                        value: option.key,
                    }))}
                    placeholder={placeholder}
                    selectedValues={dataValue}
                />
            </Label>
        </div>
    );
};

export { MultiSelectField as MultiSelectFieldBase };
export default MultiSelectField;
