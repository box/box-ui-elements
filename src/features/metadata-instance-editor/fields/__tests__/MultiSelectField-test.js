import React from 'react';

import MultiSelectField from '../MultiSelectField';

describe('features/metadata-instance-editor/fields/MultiSelectField', () => {
    test('should correctly render', () => {
        const wrapper = shallow(<MultiSelectField dataValue={['value']} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should correctly render with description', () => {
        const wrapper = shallow(<MultiSelectField dataValue={['value']} description="description" />);
        expect(wrapper).toMatchSnapshot();
    });
});
