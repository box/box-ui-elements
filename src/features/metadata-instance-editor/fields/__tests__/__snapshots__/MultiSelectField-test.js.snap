// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/metadata-instance-editor/fields/MultiSelectField should correctly render 1`] = `
<div
  className="metadata-instance-editor-field-multi-select"
>
  <Label>
    <MultiSelectField
      onChange={[Function]}
      options={Array []}
      placeholder={
        <FormattedMessage
          defaultMessage="Select Values"
          id="boxui.metadataInstanceEditor.fieldMultiSelectValue"
        />
      }
      selectedValues={
        Array [
          "value",
        ]
      }
    />
  </Label>
</div>
`;

exports[`features/metadata-instance-editor/fields/MultiSelectField should correctly render with description 1`] = `
<div
  className="metadata-instance-editor-field-multi-select"
>
  <Label>
    <i
      className="metadata-instance-editor-field-multi-select-desc"
    >
      description
    </i>
    <MultiSelectField
      onChange={[Function]}
      options={Array []}
      placeholder={
        <FormattedMessage
          defaultMessage="Select Values"
          id="boxui.metadataInstanceEditor.fieldMultiSelectValue"
        />
      }
      selectedValues={
        Array [
          "value",
        ]
      }
    />
  </Label>
</div>
`;
